File: src/main/java/br/com/g1bet/service/ApostaService.java
package br.com.g1bet.service;

import br.com.g1bet.exceptions.SaldoInsuficienteException;
import br.com.g1bet.model.Aposta;
import br.com.g1bet.model.Partida;
import br.com.g1bet.model.TipoApostaEnum;
import br.com.g1bet.model.Usuario;
import br.com.g1bet.dto.request.ApostaRequest;
import br.com.g1bet.dto.response.ApostaResponse;
import br.com.g1bet.repository.ApostaRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ApostaService {

    private final ApostaRepository repository;
    private final UsuarioService usuarioService;
    private final PartidaService partidaService;

    public ApostaService(ApostaRepository repository, UsuarioService usuarioService, PartidaService partidaService) {
        this.repository = repository;
        this.usuarioService = usuarioService;
        this.partidaService = partidaService;
    }

    public ResponseEntity<ApostaResponse> buscarId(Long id) {
        return repository.findById(id)
                .map(resp -> ResponseEntity.ok(ApostaResponse.toApostaResponse(resp)))
                .orElse(ResponseEntity.notFound().build());
    }

    public List<Aposta> buscarTipo(TipoApostaEnum tipoDeAposta) {
        return repository.findAllByTipo(tipoDeAposta);
    }

    public ApostaResponse apostar(ApostaRequest apostaRequest) {
        Usuario usuario = usuarioService.buscar(apostaRequest.getUsuario());

        if (usuario.getSaldoUsuario() < apostaRequest.getValorApostado()) {
            throw new SaldoInsuficienteException("Saldo insuficiente");
        }

        Partida partida = partidaService.findById(apostaRequest.getPartida());

        Aposta aposta = new Aposta();
        aposta.setTipo(apostaRequest.getTipo());
        aposta.setUsuario(usuario);
        aposta.setPartida(partida);
        aposta.setValorApostado(apostaRequest.getValorApostado());

        aposta = repository.save(aposta);
        return ApostaResponse.toApostaResponse(aposta);
    }

    public void deleteById(Long id) {
        repository.deleteById(id);
    }

    public List<ApostaResponse> exibirHistorico(Long id) {
        return repository.findByUsuarioId(id).stream().map(aposta -> ApostaResponse.toApostaResponse(aposta)).toList();
    }

}

File: src/main/java/br/com/g1bet/service/ApostaService.kt
package br.com.g1bet.service

import br.com.g1bet.dto.request.ApostaRequest
import br.com.g1bet.dto.response.ApostaResponse
import br.com.g1bet.mapper.exceptions.SaldoInsuficienteException
import br.com.g1bet.model.Aposta
import br.com.g1bet.model.TipoApostaEnum
import br.com.g1bet.repository.ApostaRepository
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import java.util.function.Function

@Service
class ApostaService(private val repository: ApostaRepository, private val usuarioService: UsuarioService, private val partidaService: PartidaService) {
    fun buscarId(id: Long): ResponseEntity<ApostaResponse> {
        return repository.findById(id)
                .map { resp -> ResponseEntity.ok(ApostaResponse.toApostaResponse(resp)) }
                .orElse(ResponseEntity.notFound().build())
    }

    fun buscarTipo(tipoDeAposta: TipoApostaEnum?): List<Aposta?>? {
        return repository.findAllByTipo(tipoDeAposta)
    }

    fun apostar(apostaRequest: ApostaRequest): ApostaResponse {
        val usuario = usuarioService.buscar(apostaRequest.usuario)
        if (usuario.getSaldoUsuario() < apostaRequest.valorApostado) {
            throw SaldoInsuficienteException("Saldo insuficiente")
        }
        val partida = partidaService.findById(apostaRequest.partida)
        var aposta = Aposta()
        aposta.tipo = apostaRequest.tipo
        aposta.usuario = usuario
        aposta.partida = partida
        aposta.valorApostado = apostaRequest.valorApostado
        aposta = repository.save(aposta)
        return ApostaResponse.toApostaResponse(aposta)
    }

    fun deleteById(id: Long) {
        repository.deleteById(id)
    }

    fun exibirHistorico(id: Long?): List<ApostaResponse> {
        return repository.findByUsuarioId(id)!!.stream().map<ApostaResponse>(Function { aposta: Aposta? -> ApostaResponse.toApostaResponse(aposta) }).toList()
    }
}